localstack/services/firehose/provider.py

```diff
@@ -607,29 +625,49 @@ def _preprocess_records(self, processor: Dict, records: List[Record]) -> List[Di
             result = response.get("Payload").read()
             result = json.loads(to_str(result))
             records = result.get("records", []) if result else []
+        elif proc_type == "MetadataExtraction":
+            if "MetadataExtractionQuery" in parameters:
+                extractions = self._get_metadata_extraction_queries(parameters["MetadataExtractionQuery"])
+                for record in records:
+                    if "data" in record:
+                        data = record["data"]
+                    if "Data" in record:
+                        data = record["Data"]
+                    data = json.loads(base64.b64decode(data))
+                    for extraction in extractions:
+                        if extraction["value"] in data:
+                            record["kinesisRecordMetadata"][extraction["name"]] = data[extraction["value"]]
+                        elif "default" in extraction:
+                            record["kinesisRecordMetadata"][extraction["name"]] = extraction["default"]
         else:
             LOG.warning("Unsupported Firehose processor type '%s'", proc_type)
         return records
 
+    def _replace_prefix_dynamic_partitions(self, metadata, groups):
+        return metadata[''.join(list(groups.group(1)))]
+
     def _put_records_to_s3_bucket(
         self,
         stream_name: str,
         records: List[Dict],
         s3_destination_description: S3DestinationDescription,
     ):
         bucket = s3_bucket_name(s3_destination_description["BucketARN"])
-        prefix = s3_destination_description.get("Prefix", "")
+        original_prefix = s3_destination_description.get("Prefix", "")
 
         s3 = connect_to_resource("s3")
-        batched_data = b"".join([base64.b64decode(r.get("Data") or r.get("data")) for r in records])
+        for r in records:
```

Comment:
does this undo the batching - we used to have this structure - join all the records into a batch to transmit, and transmit it

```python
        batched_data = b"".join([base64.b64decode(r.get("Data") or r.get("data")) for r in records])
        s3.Object(bucket, obj_path).put(Body=batched_data)
```

We now have this instead, for each record, build the payload and transmit it

```python
for r in records:
  ...
  batched_data = b"".join([base64.b64decode(r.get("Data") or r.get("data"))])
  s3.Object(bucket, obj_path).put(Body=batched_data)
```

I think this could lead to a large number of additional network requests (and possibly run up against rate limiting).

Reply:
oh - this is localstack! does any of that even matter?

Reply:
It does - i think i would have to do a group by on keys - let me look into that.

This is just for local dev though, so I'm not sure we'd ever do the amount of requests that would get us into rate limiting...

Reply:
@djuber I fixed it anyways!